reservation_units/querysets.py

```diff
@@ -0,0 +1,418 @@
+from copy import copy
+from datetime import date, datetime, time, timedelta
+from decimal import Decimal
+from math import ceil
+from typing import TYPE_CHECKING, Self, TypedDict
+
+from django.db import models
+from django.db.models import Case, Prefetch, Q, Value, When
+from django.utils import timezone
+from django.utils.timezone import get_default_timezone
+from elasticsearch_django.models import SearchResultsQuerySet
+
+from applications.choices import ApplicationRoundStatusChoice
+from applications.models import ApplicationRound
+from opening_hours.models import ReservableTimeSpan
+from opening_hours.utils.reservable_time_span_client import TimeSpanElement, override_reservable_with_closed_time_spans
+
+if TYPE_CHECKING:
+    from opening_hours.models import ReservationUnit
+
+DEFAULT_TIMEZONE = get_default_timezone()
+
+INTERVAL_TO_MINUTES: dict[str, int] = {
+    "interval_15_mins": 15,
+    "interval_30_mins": 30,
+    "interval_60_mins": 60,
+    "interval_90_mins": 90,
+}
+
+
+def _get_hard_closed_time_spans_for_reservation_unit(reservation_unit: "ReservationUnit") -> list[TimeSpanElement]:
+    reservation_unit_closed_time_spans: list[TimeSpanElement] = []
+
+    if reservation_unit.reservation_begins:
+        reservation_unit_closed_time_spans.append(
+            TimeSpanElement(
+                start_datetime=datetime.min.replace(tzinfo=DEFAULT_TIMEZONE),
+                end_datetime=reservation_unit.reservation_begins,
+                is_reservable=False,
+            )
+        )
+    if reservation_unit.reservation_ends:
+        reservation_unit_closed_time_spans.append(
+            TimeSpanElement(
+                start_datetime=reservation_unit.reservation_ends,
+                end_datetime=datetime.max.replace(tzinfo=DEFAULT_TIMEZONE),
+                is_reservable=False,
+            )
+        )
+
+    if reservation_unit.publish_ends:
```

Comment:
Need to exclude the case where publishing has restarted (or set TimeSpanElement.end_datetime in that situation accordingly, instead of datetime.max): `publish_ends < publish_begins <= now`, cf. https://github.com/City-of-Helsinki/tilavarauspalvelu-core/blob/d795967c6fadff4e6d7a2cc1c00889c32f5a17e4/reservation_units/utils/reservation_unit_state_helper.py#L175